core/src/main/java/com/linecorp/armeria/common/logging/TracingContext.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2024 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * A holder class which has the tracing(e.g. zipkin) information.
+ */
+public final class TracingContext {
```

Comment:
so here's my TL;DR; info. If this happens, this should be an interface, which will allow lazy loading of details only when present, and also a no-op impl. don't worry about random properties, and focus on traceID/spanId and doing it here vs alternates like via MDC (status quo for most)

Here are the edge cases to consider:

1. there is no trace at all, so should TracingContext be returned as null in this case? or nullable traceId and spanId
2. documentation about this type about which span it is about. For example, is this the server request? or any request?
3. use cases this supports, exactly. For example, is this for setting MDC? If so, when would you use this vs normal MDC integration
4. how to disable this, and what is the fallback (e.g maybe behavior of unsampled is same as disabled)

Tests should cover these questions, and a benchmark to show we aren't accidentally copying data per request when there is no trace or this feature is disabled.